File: ScorpioCommons/Java/ScorpioCommons/src/Scorpio/Message/IMessage.java
package Scorpio.Message;

import Scorpio.Commons.ScorpioUtil;
import Scorpio.Commons.ScorpioWriter;

public abstract class IMessage {
    protected int __Sign = 0;
    protected final void AddSign(int index) {
        __Sign = ScorpioUtil.AddSign(__Sign, index);
    }
    public final boolean HasSign(int index) {
        return ScorpioUtil.HasSign(__Sign, index);
    }
    public final byte[] Serialize() {
    	ScorpioWriter writer = new ScorpioWriter();
        Write(writer);
        return writer.ToArray();
    }
    public abstract void Write(ScorpioWriter writer);
}

File: ScorpioProto-CPP/Scorpio/Message/IMessage.cpp
#include "IMessage.h"
#include "../Commons/ScorpioUtil.h"
using namespace Scorpio::Commons;
namespace Scorpio {
	namespace Message {
		void IMessage::AddSign(int index) {
			__Sign = ScorpioUtil::AddSign(__Sign, index);
		}
		bool IMessage::HasSign(int index) {
			return ScorpioUtil::HasSign(__Sign, index);
		}
		char * IMessage::Serialize() {
			ScorpioWriter * writer = new ScorpioWriter();
			Write(writer);
			char * ret = writer->ToArray();
			delete writer;
			writer = nullptr;
			return ret;
		}
	}
}